core/utils/utils.go

```diff
@@ -418,46 +418,48 @@ func WaitGroupChan(wg *sync.WaitGroup) <-chan struct{} {
 // channel is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (ctx context.Context, cancel context.CancelFunc) {
-	ctx, cancel = context.WithCancel(parentCtx)
-
-	go func() {
-		select {
-		case <-chStop:
-		case <-ctx.Done():
-		}
-		cancel()
-	}()
-
-	return ctx, cancel
+func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithCancel(parentCtx))
 }
 
 // ContextFromChan creates a context that finishes when the provided channel
 // receives or is closed.
-// When channel closes, the ctx.Err() will always be context.Canceled
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
 func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		select {
-		case <-chStop:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	return ctx, cancel
+	return StopChan(chStop).NewCtx()
 }
 
-// ContextFromChanWithDeadline creates a context with a deadline that finishes when the provided channel
+// ContextFromChanWithTimeout creates a context with a timeout that finishes when the provided channel
 // receives or is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func ContextFromChanWithDeadline(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func ContextFromChanWithTimeout(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithTimeout(context.Background(), timeout))
+}
+
+type StopChan <-chan struct{}
+
+func NewStopChan() (ch StopChan, stop func()) {
+	c := make(chan struct{})
+	return c, func() { close(c) }
+}
+
+// NewCtx returns a new background [context.Context] that will be cancelled when the StopChan is closed.
+//
+// Note: The [context.CancelFunc] must be called to prevent memory leaks.
+func (s StopChan) NewCtx() (context.Context, context.CancelFunc) {
+	return s.CancelCtx(context.WithCancel(context.Background()))
+}
+
+// CancelCtx spawns a goroutine to cancel a [context.Context] when StopChan is closed.
+// Returns ctx and cancel unmodified, for convenience.
+//
+// Note: The [context.CancelFunc] must be called to prevent memory leaks.
```

Comment:
the note is wrong. either
- the stop fn must be called
- or the returned cancel fn must be called
to prevent leaks

Reply:
Hmmm I was trying to carry forward the note from the existing helpers :thinking: 
I think the intention is to inform that the author of the calling code must also write code that calls cancel - not that it must execute or be the ultimate trigger. Is there a better way to say this? Or are we just adding to the confusion, and would it be more clear to just leave it out under the assumption that it should be used normally?

Reply:
Removed for now